File: common/src/main/java/com/pluxity/user/repository/UserRepository.java
package com.pluxity.user.repository;

import com.pluxity.user.entity.User;
import java.util.List;
import java.util.Optional;
import org.springframework.data.jpa.repository.EntityGraph;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.lang.NonNull;

public interface UserRepository extends JpaRepository<User, Long> {
    @EntityGraph(
            attributePaths = {
                "userRoles",
                "userRoles.role",
                "userRoles.role.rolePermissions",
                "userRoles.role.rolePermissions.permission"
            })
    @NonNull
    List<User> findAll();

    @EntityGraph(
            attributePaths = {
                "userRoles",
                "userRoles.role",
                "userRoles.role.rolePermissions",
                "userRoles.role.rolePermissions.permission"
            })
    @NonNull
    Optional<User> findById(@NonNull Long id);

    @EntityGraph(
            attributePaths = {
                "userRoles",
                "userRoles.role",
                "userRoles.role.rolePermissions",
                "userRoles.role.rolePermissions.permission"
            })
    Optional<User> findByUsername(@NonNull String username);
}

File: common/src/main/kotlin/com/pluxity/user/repository/UserRepository.kt
package com.pluxity.user.repository

import com.pluxity.user.entity.User
import org.springframework.data.domain.Sort
import org.springframework.data.jpa.repository.EntityGraph
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query

interface UserRepository : JpaRepository<User, Long> {
    @Query(
        """
        SELECT u.username FROM User u
        JOIN u.userRoles ur
        JOIN ur.role r
        WHERE r.name = :roleName
        """,
    )
    fun findUsernamesByRoleName(roleName: String): List<String>

    @EntityGraph(
        attributePaths = [
            "userRoles", "userRoles.role",
        ],
    )
    fun findAllBy(sort: Sort): List<User>

    @EntityGraph(
        attributePaths = [
            "userRoles.user",
            "userRoles.role.rolePermissions.permission.resourcePermissions",
            "userRoles.role.rolePermissions.permission.domainPermissions",
        ],
    )
    fun findWithGraphById(id: Long): User?

    @EntityGraph(
        attributePaths = [
            "userRoles", "userRoles.role",
        ],
    )
    fun findByUsername(username: String): User?
}
